src/allocation_engine/interval_tree.rs

```diff
@@ -37,10 +41,125 @@ pub struct Constraint {
 /// memory slot that will be assigned to a device by the VMM.
 /// This structure represents the key of the Node object in
 /// the interval tree implementation.
-#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Hash, Ord)]
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Hash, Ord, Debug)]
 pub struct Range {
-    pub min: u64,
-    pub max: u64,
+    /// Lower boundary of the interval.
+    min: u64,
+    /// Upper boundary of the interval.
+    max: u64,
+}
+
+impl Range {
+    /// Create a new Range object.
+    pub fn new(min: u64, max: u64) -> Result<Self> {
+        if min > max {
+            return Err(Error::InvalidRange(min, max));
+        }
+        Ok(Range { min, max })
+    }
+
+    /// Get length of the range.
+    pub fn length(&self) -> u64 {
+        self.max - self.min + 1
+    }
+
+    /// Check whether two Range objects intersect with each other.
+    pub fn intersect(&self, other: &Range) -> bool {
+        max(self.min, other.min) <= min(self.max, other.max)
+    }
+
+    /// Check whether the key is fully covered.
+    pub fn contain(&self, other: &Range) -> bool {
+        self.min <= other.min && self.max >= other.max
+    }
+
+    /// Create a new Range object with min being the first value from the range
+    /// that is aligned to `align`.
+    ///
+    /// # Examples
+    /// ```rust
+    /// extern crate vm_allocator;
+    /// use vm_allocator::allocation_engine::Range;
+    /// use vm_allocator::Result;
+    ///
+    /// fn intervals_align_forward() -> Result<Range> {
+    ///     let range_to_align = Range::new(2, 6)?;
+    ///     let res = range_to_align.align_forward(2)?;
+    ///     assert_eq!(res.length(), 5);
+    ///     Ok(res)
+    /// }
+    ///
+    /// intervals_align_forward().unwrap();
```

Comment:
I've read this doc comments a few times and I still don't understand what this method does.

I think it would be helpful to keep the example in the doc, but show only three lines (and not have the `intervals_align_forward` function at all) to clearly demonstrate how the aligned range looks like
```rust
let range = Range::new(2, 6).unwrap();
let aligned_range = range.align_forward(2).unwrap();
assert_eq!(aligned_range, Range::new(X, Y).unwrap());
```